README.md

```diff
@@ -513,6 +513,32 @@ and [development](https://docs.ansible.com/ansible/latest/dev_guide/index.html).
   described in the 
   [Ansible docs](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html).
 
+## Collections
+
+### Modules and module_utils
+In a collection, the names of the module and module_utils files must be unique
+since all of them are located in
+`COLLECTION_NAMESPACE/COLLECTION_NAME/plugins/modules` (modules) and
+`COLLECTION_NAMESPACE/COLLECTION_NAME/plugins/module_utils` (module supporting
+files).  It is recommended to:
+* start module filenames with `ROLENAME_`
+* use a sub-directory `module_utils/ROLENAME/` for the role-specific
+  module_utils files
+* avoid generic names such as `size.py`
+
+If you are converting several roles into collection format and bundling them in
+a collection, you may need to rename files so that they are unique.  You may
+have to edit your module files to adjust the python imports.
```

Comment:
Not sure I get your point

Reply:
You're saying that it's possible you import roles A, B, C and all have a module named `foo`. You're going to need to rename at least some of them when they join the collection. That's true.

However, this is a great argument against your point above asking for them all to be named A_foo, B_foo, C_foo. If they're all named `foo` that's probably because there's some shared functionality or similarity at least that needs to be shared more broadly than just within a particular role. So either they need to be named differently (e.g. "size" is very ambiguous) or they should NOT be named after their original role because this is clearly functionality that is needed for more than just that one specific role. So just name it `foo` and make sure it covers all the use cases in the roles.

Reply:
@greg-hellings @richm I believe it was @t-woerner who argued for such kind of disambiguation using prefixes. IIRC the reason was that as soon as someone uses a `collections:` keyword with our collection, everything gets imported into the namespace and we are back at the old problem with having modules or other plugins without any disambiguation possibly conflicting with each other.
IIRC also similar argument holds for module_utils, but this time the problem occurs even without using the `collections:` keyword. module_utils live in the same namespace, so there is a risk that we will shadow a globally provided module_util with our one. This can be prevented by creating subdirectories under module_utils and putting the actual Python files there. (That said, there is no way to guarantee that the name of subdirectory will not clash with any common module_util, right?)
@t-woerner, can you please confirm that I represented your concerns accurately?

Reply:
As with any other place that has namespaces, that collision is a problem for the consumer. Not the producer. If I do `from foo import *` and `from bar import *` and they both have component `baz` then, as the invoker, I call `foo.baz` or `bar.baz` from the point of invocation. I don't go to the authors of foo and bar and demand that they rename to `foo_baz` and `bar_baz`. That's the whole point of namespaces: to allow name collision resolution. It is going against the whole purpose of collection namespacing to further namespace the module name by appending the role that it is used in to the filename. Besides, what is to stop `oasis_roles.system` and `foo.bar` collections from both having role named `do_thing` and thus they both end up with module `do_thing_module`.

You're not back the same problem as before. You now have the FQCN to access the desired module through. Don't further complicate naming schemes with unnecessary bits that are ultimately unproductive and, if you're using the module in more than one place, misleading.

The same does NOT hold for module_utils. They are not shoveled into a flat namespace with all other module_utils added. They would be accessed at `from ansible_collections.oasis_roles.system.plugins.module_utils.some_util_file import my_util_function`. The `oasis_roles.system` is the namespacing provided by my collection. At least this is [how it is documented](https://docs.ansible.com/ansible/devel/dev_guide/migrating_roles.html#update-module-utils-role). So, again, introducing additional namespacing and folder structure under here is possible, if a collection has supremely complex module_utils. But do not force it a priori in an attempt to avoid naming collisions with globally available module_utils or with module_utils that live in another collection.

This whole proposal is attempting to solve a problem which was important back in the pre-collection days but which is the exact problem that collections were created to solve. Don't re-invent the wheel on top of collections again. Just use the collection.

Reply:
> @greg-hellings @richm I believe it was @t-woerner who argued for such kind of disambiguation using prefixes. IIRC the reason was that as soon as someone uses a `collections:` keyword with our collection, everything gets imported into the namespace and we are back at the old problem with having modules or other plugins without any disambiguation possibly conflicting with each other.
> IIRC also similar argument holds for module_utils, but this time the problem occurs even without using the `collections:` keyword. module_utils live in the same namespace, so there is a risk that we will shadow a globally provided module_util with our one. This can be prevented by creating subdirectories under module_utils and putting the actual Python files there. (That said, there is no way to guarantee that the name of subdirectory will not clash with any common module_util, right?)
> @t-woerner, can you please confirm that I represented your concerns accurately?

Yes, that is more or less it. You can still use the full qualified name, but we should make sure that an addition or change within system roles will not easily introduce a conflict with other collections that customers are using. Especially with simple and common names.

Reply:
@t-woerner assuming that this was about the first point (module/plugin names), the second (module_utils placement) seems to be sufficiently resolved. As @greg-hellings says, module/plugin names issue is addressed by the collection structure. You seem to be concerned about using the `collections:` keyword, which is very much like `from ... import *` in Python and brings similar problems. To which the response shall be: don't use the `collections:` keyword then, just as it is better to not use `from ... import *`. Isn't that enough?